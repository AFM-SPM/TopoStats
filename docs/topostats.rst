topostats package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   topostats.logs
   topostats.tracing

Submodules
----------

.. toctree::
   :maxdepth: 4

   topostats.entry_point
   topostats.filters
   topostats.grains
   topostats.grainstats
   topostats.io
   topostats.plotting
   topostats.plottingfuncs
   topostats.processing
   topostats.run_topostats
   topostats.scars
   topostats.statistics
   topostats.theme
   topostats.thresholds
   topostats.utils
   topostats.validation

Module contents
---------------

.. automodule:: topostats
   :members:
   :undoc-members:
   :show-inheritance:
