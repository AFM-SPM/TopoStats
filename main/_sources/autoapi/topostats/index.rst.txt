topostats
=========

.. py:module:: topostats

.. autoapi-nested-parse::

   
   Topostats.
















   ..
       !! processed by numpydoc !!


Submodules
----------

.. toctree::
   :maxdepth: 1

   /autoapi/topostats/array_manipulation/index
   /autoapi/topostats/config/index
   /autoapi/topostats/entry_point/index
   /autoapi/topostats/filters/index
   /autoapi/topostats/grains/index
   /autoapi/topostats/grainstats/index
   /autoapi/topostats/io/index
   /autoapi/topostats/logs/index
   /autoapi/topostats/plotting/index
   /autoapi/topostats/plottingfuncs/index
   /autoapi/topostats/processing/index
   /autoapi/topostats/run_modules/index
   /autoapi/topostats/scars/index
   /autoapi/topostats/statistics/index
   /autoapi/topostats/theme/index
   /autoapi/topostats/thresholds/index
   /autoapi/topostats/tracing/index
   /autoapi/topostats/unet_masking/index
   /autoapi/topostats/utils/index
   /autoapi/topostats/validation/index


Attributes
----------

.. autoapisummary::

   topostats.LOGGER
   topostats.__version__
   topostats.__release__
   topostats.TOPOSTATS_DETAILS
   topostats.TOPOSTATS_VERSION
   topostats.TOPOSTATS_COMMIT
   topostats.CONFIG_DOCUMENTATION_REFERENCE


Classes
-------

.. autoapisummary::

   topostats.ImageGrainCrops
   topostats.Colormap
   topostats.TopoStats


Functions
---------

.. autoapisummary::

   topostats.setup_logger
   topostats.log_topostats_version


Package Contents
----------------

.. py:class:: ImageGrainCrops

   
   Dataclass for storing the crops of grains in an image.

   .. attribute:: above

      Grains in the above direction.

      :type: GrainCropDirection | None

   .. attribute:: below

      Grains in the below direction.

      :type: GrainCropDirection | None















   ..
       !! processed by numpydoc !!

   .. py:attribute:: above
      :type:  GrainCropsDirection | None


   .. py:attribute:: below
      :type:  GrainCropsDirection | None


   .. py:method:: __eq__(other: object) -> bool

      
      Check if two ImageGrainCrops objects are equal.

      :param other: Object to compare to.
      :type other: object

      :returns: True if the objects are equal, False otherwise.
      :rtype: bool















      ..
          !! processed by numpydoc !!


   .. py:method:: image_grain_crops_to_dict() -> dict[str, numpy.typing.NDArray[numpy.bool_] | dict[str:Any]]

      
      Convert ImageGrainCrops to dictionary indexed by attributes.

      :returns: Dictionary indexed by attribute of the grain attributes.
      :rtype: dict[str, Any]















      ..
          !! processed by numpydoc !!


   .. py:method:: debug_locate_difference(other: object) -> None

      
      Debug function to find the culprit when two ImageGrainCrops objects are not equal.

      :param other: Object to compare to.
      :type other: object

      :raises ValueError: If the objects are not equal.















      ..
          !! processed by numpydoc !!


.. py:function:: setup_logger(log_name: str = LOGGER_NAME) -> logging.Logger

   
   Logger setup.

   The logger for the module is initialised when the module is loaded (as this functions is called from
   __init__.py). This creates two stream handlers, one for general output and one for errors which are formatted
   differently (there is greater information in the error formatter). To use in modules import the 'LOGGER_NAME' and
   create a logger as shown in the Examples, it will inherit the formatting and direction of messages to the correct
   stream.

   :param log_name: Name under which logging information occurs.
   :type log_name: str

   :returns: Logger object.
   :rtype: logging.Logger

   .. rubric:: Examples

   To use the logger in (sub-)modules have the following.

       import logging
       from topostats.logs.logs import LOGGER_NAME

       LOGGER = logging.getLogger(LOGGER_NAME)

       LOGGER.info('This is a log message.')















   ..
       !! processed by numpydoc !!

.. py:class:: Colormap(name: str = 'nanoscope')

   
   Class for setting the Colormap.

   :param name: Name of colormap to use.
   :type name: str















   ..
       !! processed by numpydoc !!

   .. py:attribute:: name
      :value: 'nanoscope'



   .. py:attribute:: cmap
      :value: None



   .. py:method:: __str__() -> str

      
      Return string representation of object.

      :returns: String detailing the colormap.
      :rtype: str















      ..
          !! processed by numpydoc !!


   .. py:method:: set_cmap(name: str) -> None

      
      Set the ColorMap.

      :param name: Name of the colormap to return.
      :type name: str















      ..
          !! processed by numpydoc !!


   .. py:method:: get_cmap() -> matplotlib.cm

      
      Return the matplotlib.cm colormap object.

      :returns: Matplotlib Color map object.
      :rtype: matplotlib.cm















      ..
          !! processed by numpydoc !!


   .. py:method:: nanoscope() -> matplotlib.colors.LinearSegmentedColormap
      :staticmethod:


      
      Matplotlib compatible colormap that replicates the Bruker Nanoscope colorscale.

      The colormap is implemented in Gwyddion's GwyGradient via 'Nanoscope.txt'.

      :returns: MatplotLib LinearSegmentedColourmap that replicates Bruker Nanoscope colorscale.
      :rtype: LinearSegmentedColormap















      ..
          !! processed by numpydoc !!


   .. py:method:: gwyddion() -> matplotlib.colors.LinearSegmentedColormap
      :staticmethod:


      
      Set RGBA colour map for the Gwyddion.net colour gradient.

      :returns: The 'gwyddion' colormap.
      :rtype: LinearSegmentedColormap















      ..
          !! processed by numpydoc !!


   .. py:method:: blue() -> matplotlib.colors.ListedColormap
      :staticmethod:


      
      Set RGBA colour map of just the colour blue.

      :returns: The 'blue' colormap.
      :rtype: ListedColormap















      ..
          !! processed by numpydoc !!


   .. py:method:: blue_purple_green() -> matplotlib.colors.ListedColormap
      :staticmethod:


      
      RGBA colour map of just the colour blue/purple/green.

      :returns: The 'blue/purple/green' colormap.
      :rtype: ListedColormap















      ..
          !! processed by numpydoc !!


.. py:data:: LOGGER

.. py:data:: __version__

.. py:data:: __release__
   :value: ''


.. py:data:: TOPOSTATS_DETAILS

.. py:data:: TOPOSTATS_VERSION

.. py:data:: TOPOSTATS_COMMIT

.. py:data:: CONFIG_DOCUMENTATION_REFERENCE
   :value: Multiline-String

   .. raw:: html

      <details><summary>Show Value</summary>

   .. code-block:: python

      """# For more information on configuration and how to use it:
      # https://afm-spm.github.io/TopoStats/main/configuration.html
      """

   .. raw:: html

      </details>



.. py:class:: TopoStats

   
   Class for storing TopoStats objects.

   .. attribute:: image_grain_crops

      ImageGrainCrops of processed image.

      :type: ImageGrainCrops | None

   .. attribute:: filename

      Filename.

      :type: str | None

   .. attribute:: pixel_to_nm_scaling

      Pixel to nanometre scaling.

      :type: str | None

   .. attribute:: img_path

      Original path to image.

      :type: str | None

   .. attribute:: image

      Flattened image (post ''Filter()'').

      :type: npt.NDArray | None

   .. attribute:: image_original

      Original image.

      :type: npt.NDArray | None

   .. attribute:: topostats_version

      TopoStats version.

      :type: str | None















   ..
       !! processed by numpydoc !!

   .. py:property:: image_grain_crops
      :type: grains.ImageGrainCrops


      
      Getter for the Image Grain Crops.

      :returns: Image Grain Crops.
      :rtype: ImageGrainCrops















      ..
          !! processed by numpydoc !!


   .. py:property:: filename
      :type: str


      
      Getter for the ''filename'' attribute.

      :returns: Image filename.
      :rtype: str















      ..
          !! processed by numpydoc !!


   .. py:property:: pixel_to_nm_scaling
      :type: str


      
      Getter for the ''pixel_to_nm_scaling'' attribute.

      :returns: Image ''pixel_to_nm_scaling''.
      :rtype: str















      ..
          !! processed by numpydoc !!


   .. py:property:: img_path
      :type: pathlib.Path


      
      Getter for the ''img_path'' attribute.

      :returns: Path to original image on disk.
      :rtype: Path















      ..
          !! processed by numpydoc !!


   .. py:property:: image
      :type: str


      
      Getter for the ''image'' attribute, post filtering.

      :returns: Image image.
      :rtype: str















      ..
          !! processed by numpydoc !!


   .. py:property:: image_original
      :type: str


      
      Getter for the ''image_original'' attribute.

      :returns: Original image.
      :rtype: str















      ..
          !! processed by numpydoc !!


   .. py:property:: topostats_version
      :type: str


      
      Getter for the ''topostats_version'' attribute, post filtering.

      :returns: Version of TopoStats the class was created with.
      :rtype: str















      ..
          !! processed by numpydoc !!


   .. py:method:: __eq__(other: object) -> bool

      
      Check if two TopoStats objects are equal.

      :param other: Object to compare to.
      :type other: object

      :returns: True if the objects are equal, False otherwise.
      :rtype: bool















      ..
          !! processed by numpydoc !!


   .. py:method:: topostats_to_dict() -> dict[str, str | grains.ImageGrainCrops | numpy.typing.NDArray]

      
      Convert ''TopoStats'' object to dictionary.

      :returns: Dictionary of ''TopoStats'' object.
      :rtype: dict[str, str | ImageGrainCrops | npt.NDArray]















      ..
          !! processed by numpydoc !!


.. py:function:: log_topostats_version() -> None

   
   Log the TopoStats version, commit and date to system logger.
















   ..
       !! processed by numpydoc !!

