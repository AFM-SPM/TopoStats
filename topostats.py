"""
File for tracking hard-coded variables (for now)
"""

import argparse

### List of hard-coded variables in pywgwytracing.py

# global variables
gwy_colobar_ln48 = "/module/pixmap/ztype"
gwy_linematchmethod_ln51= '/module/linematch/method'
gwy_linematchmaxdegree_ln52 = "/module/linematch/max_degree"
gwy_polylevel_coldegree_ln53 = "/module/polylevel/col_degree"

# heightediting()
datafield_grains_mark_height_ln128 = int()
gwy_polylevel_masking_ln131 = int()
gwy_linematch_masking_ln135 = int()
filterwidth_ln144 = float()

# editfile()
datafield_markheight_ln170 = int()
gwy_polylevel_masking_ln173 = int()
gwy_linematch_masking_ln177 = gwy_linematch_masking_ln135
gwy_polylevel_masking_ln181 = int()
datafield_gaussfilter_ln195 = int()

# grainfinding()
grain_gauss_ln211 = float()
datafield_mask_outliers_ln220 = float()

# remove_large_objects()
datafield_mask_outliers_ln257 = float()

# remove_small_objects()
datafield_mask_outliers_ln283 = float()

# grain_thinning()
gaussian_size_ln488 = float()

# savefiles()
savefile_type_ln535 = ["/module/pixmap/xytype"]

# saveunknownfiles()
savefile_type_ln573 = ["/module/pixmap/xytype"]

# savecroppedfiles()
savefile_type_ln645 = ["/module/pixmap/xytype"]

# __main__()
path_ln709 = str()
fileend_ln712 = []
filetype_ln713 = str()
extension_ln715 = str()
minheightscale_ln717 = float()
maxheightscale_ln718 = float()
minarea_ln722 = float()
maxdeviation_ln726 = float()
mindeviation_ln727 = float()
cropwidth_ln733 = float()
splitwidth_ln734 = float()
bins_ln736 = int()
channels_ln755 = list()


### dnatracing.py

# getNumpyArraysfromGwyddion()
sigma_ln94 = float()

# getDisorderedTrace()
sigma_ln109 = float()

# purgeObviousCrap()
maxlen_ln126 = int()

# getFittedTraces()
index_width_ln212 = int()

# getSplinedTraces()
step_size_ln376 = float()
interp_size_ln377 = float()

### tracingfuncs.py

# I didn't find any hard-coded variables so far

### statsplotting.py
### TFOplot.py
### traceplotting.py

# not checking these for now due to ongoing merging project
