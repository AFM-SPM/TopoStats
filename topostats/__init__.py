"""Topostats"""
from importlib.metadata import version

from .logs.logs import setup_logger

LOGGER = setup_logger()

release = version("topostats")
__version__ = ".".join(release.split("."[:2]))
