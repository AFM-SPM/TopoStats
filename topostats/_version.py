# file generated by setuptools_scm
# don't change, don't track in version control
__version__ = version = "2.0.1.dev369+g21461a8.d20230301"
__version_tuple__ = version_tuple = (2, 0, 1, "dev369", "g21461a8.d20230301")
