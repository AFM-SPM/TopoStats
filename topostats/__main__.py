import topostats.topotracing as topotracing

topotracing.main()
