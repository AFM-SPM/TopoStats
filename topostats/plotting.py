"""Plotting and summary of Statistics"""
import argparse as arg
from collections import defaultdict
import importlib.resources as pkg_resources
import logging
from pathlib import Path
import sys
from typing import Union, Dict
import yaml

import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

from topostats.io import read_yaml, save_pkl, write_yaml
from topostats.logs.logs import LOGGER_NAME
from topostats.utils import update_config

LOGGER = logging.getLogger(LOGGER_NAME)


# pylint: disable=too-many-arguments
# pylint: disable=too-many-instance-attributes
# pylint: disable=too-many-locals


def create_parser() -> arg.ArgumentParser:
    """Create a parser for reading options."""
    parser = arg.ArgumentParser(
        description="Summarise and plot histograms, kernel density estimates and scatter plots of TopoStats"
        "grain and DNA Tracing statistics."
    )
    parser.add_argument("-i", "--input_csv", dest="csv_file", required=False, help="Path to CSV file to plot.")
    parser.add_argument(
        "-c",
        "--config_file",
        dest="config_file",
        required=False,
        help="Path to a YAML configuration file.",
    )
    parser.add_argument(
        "-l",
        "--var_to_label",
        dest="var_to_label",
        required=False,
        help="Path to a YAML plotting dictionary that maps variable names to labels.",
    )
    parser.add_argument(
        "--create-config-file",
        dest="create_config_file",
        type=str,
        required=False,
        help="Filename to write a sample YAML configuration file to (should end in '.yaml').",
    )
    parser.add_argument(
        "--create-label-file",
        dest="create_label_file",
        type=str,
        required=False,
        help="Filename to write a sample YAML label file to (should end in '.yaml').",
    )
    return parser


class TopoSum:
    """Class for summarising grain statistics in plots."""

    def __init__(
        self,
        df: Union[pd.DataFrame] = None,
        base_dir: Union[str, Path] = None,
        csv_file: Union[str, Path] = None,
        stat_to_sum: str = None,
        molecule_id: str = "molecule_number",
        image_id: str = "image",
        hist: bool = True,
        stat: str = "count",
        bins: int = 12,
        kde: bool = True,
        cut: float = 20,
        figsize: tuple = (16, 9),
        alpha: float = 0.5,
        palette: str = "deep",
        file_ext: str = "png",
        output_dir: Union[str, Path] = ".",
        var_to_label: dict = None,
    ) -> None:
        """Initialise the class.

        Parameters
        ==========
        df: Union[pd.DataFrame]
            Pandas data frame of data to be summarised.
        base_dir: Union[str, Path]
            Base directory from which all paths are relative to.
        csv_file: Union[str, Path]
            CSV file of data to be summarised.
        stat_to_sum: str
            Variable to summarise.
        molecule_id: str
            Variable that uniquely identifies molecules.
        image_id: str
            Variable that uniquely identifies images.
        hist: bool
            Whether to plot histograms.
        stat: str
            Statistic to plot on histogram 'count' (default), 'freq'.
        bins: int
            Number of bins to plot.
        kde: bool
            Whether to include a Kernel Density Estimate.
        cut: float = 20,
            Cut point for KDE.
        figsize: tuple
            Figure dimensions.
        alpha: float
            Opacity to use in plots.
        palette: str = "deep"
            Seaborn colour plot to use.
        file_ext: str
            File type to save plots as 'png' (default), 'pdf', 'svg'.
        output_dir: Union[str, Path]
            Location to save plots to.
        var_to_label: dict
            Variable to label dictionary for automatically adding titles to plots.

        Returns
        =======
        """
        self.df = df if df is not None else pd.read_csv(csv_file)
        self.base_dir = base_dir
        self.stat_to_sum = stat_to_sum
        self.molecule_id = molecule_id
        self.image_id = image_id
        self.hist = hist
        self.bins = bins
        self.stat = stat
        self.kde = kde
        self.cut = cut
        self.figsize = figsize
        self.alpha = alpha
        self.palette = palette
        self.file_ext = file_ext
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(parents=True, exist_ok=True)
        self.var_to_label = var_to_label
        self.melted_data = None
        self.summary_data = None
        self.label = None
        self.sns_melt_data()
        self.set_palette()
        self._set_label(self.stat_to_sum)

    def _setup_figure(self):
        """Setup Matplotlib figure and axes."""
        fig, ax = plt.subplots(1, 1, figsize=self.figsize)
        return fig, ax

    def _outfile(self, plot_suffix: str) -> str:
        """Generate the output file name with the appropriate suffix.

        Parameters
        ----------
        plot_suffix: str
            The suffix to append to the output file.

        Returns
        -------
        str:
            Concanenated string of the outfile and plot_suffix.
        """
        return f"{self.stat_to_sum}_{plot_suffix}"

    def sns_plot(self) -> None:
        """Plot the distribution of one or more statistics as either histogram, kernel density estimates or both. Uses
        base Seaborn."""
        fig, ax = self._setup_figure()
        if self.hist and not self.kde:
            outfile = self._outfile("hist")
            sns.histplot(data=self.melted_data, x="value", bins=self.bins, stat=self.stat, hue=self.image_id)
        if self.kde and not self.hist:
            outfile = self._outfile("kde")
            sns.kdeplot(data=self.melted_data, x="value", hue=self.image_id)
        if self.hist and self.kde:
            outfile = self._outfile("hist_kde")
            sns.histplot(
                data=self.melted_data,
                x="value",
                bins=self.bins,
                stat=self.stat,
                hue=self.image_id,
                kde=True,
                kde_kws={"cut": self.cut},
            )
        plt.ticklabel_format(axis="both", style="sci", scilimits=(-3, 3))
        plt.title(self.label)
        self.set_xlim()
        self.save_plot(outfile)
        return fig, ax

    def sns_violinplot(self) -> None:
        """Violin plot of data."""
        fig, ax = self._setup_figure()
        sns.violinplot(data=self.melted_data, x=self.image_id, y="value", hue=self.image_id, alpha=self.alpha)
        plt.title(self.label)
        plt.xlabel("Image")
        plt.ylabel(self.label)
        outfile = self._outfile("violin")
        self.save_plot(outfile)
        return fig, ax

    # def sns_jointplot(self, var1: str, var2: str) -> None:
    #     """Joint distribution of two variables."""
    #     fig, ax = self._setup_figure()
    #     sns.jointplot(data=self.df, x=var1, y=var2, kind="reg")
    #     outfile = f"{'_'.join(self.stats_to_sum.keys())}_jointplot"
    #     # outfile = self._outfile("jointplot")
    #     self.save_plot(outfile)
    #     return fig, ax

    def sns_melt_data(self) -> pd.DataFrame:
        """Melt a dataframe into long format for plotting with Seaborn."""
        self.melted_data = pd.melt(
            self.df.reset_index(), id_vars=[self.molecule_id, self.image_id], value_vars=self.stat_to_sum
        )
        self.melted_data["variable"] = self.melted_data["variable"].map(self.var_to_label)
        self.melted_data.rename({"image": "Image"}, axis=1, inplace=True)
        self.image_id = "Image"
        LOGGER.info("[plotting] Data has been melted to long format for plotting.")

    def set_xlim(self, percent: float = 0.1) -> None:
        """Set the range of the x-axis.

        Parameters
        ----------
        percent: float
            Percentage of the observed range by which to extend the x-axis. Only used if supplied range is outside the
        observed values.
        """
        range_percent = percent * (self.melted_data["value"].max() - self.melted_data["value"].min())
        range_min = self.melted_data["value"].min()
        range_max = self.melted_data["value"].max()
        plt.xlim(range_min - range_percent, range_max + range_percent)
        LOGGER.info(f"[plotting] Setting x-axis range       : {range_min} - {range_max}")

    def set_palette(self):
        """Set the color palette."""
        sns.set_palette(self.palette)
        LOGGER.info(f"[plotting] Seaborn color palette : {self.palette}")

    def save_plot(self, outfile: Path) -> None:
        """Save the plot to the output_dir

        Parameters
        ----------
        outfile: str
            Output file name to save figure to.
        """
        plt.savefig(self.output_dir / f"{outfile}.{self.file_ext}")
        LOGGER.info(
            f"[plotting] Plotted {self.stat_to_sum} to : " f"{str(self.output_dir / f'{outfile}.{self.file_ext}')}"
        )

    def _set_label(self, var: str):
        """Get the label based on the column name(s).

        Parameters
        ----------
        var: str
            The variable for which a label is required.
        """
        self.label = self.var_to_label[var]
        LOGGER.debug(f"[plotting] self.label     : {self.label}")

    def summarise_by_image(self):
        """Summarise statistics by image."""
        self.summary_data = self.df.groupby(["image", "threshold"]).describe()


def toposum(config: dict) -> Dict:
    """Process plotting and summarisation of data.

    Parameters
    ----------
    config: dict
        Dictionary of summarisation options.

    Returns
    -------
    Dict
        Dictionary of nested dictionaries. Each variable has its own dictionary with keys 'dist' and 'violin' which
        contain distribution like plots and violin plots respectively (if the later are required). Each 'dist' and
       'violin' is itself a dictionary with two elements 'figures' and 'axes' which correspond to MatplotLib 'fig' and
       'ax' for that plot.
    """
    if "df" not in config.keys():
        config["df"] = pd.read_csv(config["csv_file"])
    if config["df"].isna().values.all():
        LOGGER.info("[plotting] No statistics in DataFrame. Exiting...")
        return None
    violin = config.pop("violin")
    all_stats_to_sum = config.pop("stats_to_sum")
    pickle_plots = config.pop("pickle_plots")
    figures = defaultdict()

    # Plot each variable on its own graph
    for var in all_stats_to_sum:
        if var in config["df"].columns:
            topo_sum = TopoSum(stat_to_sum=var, **config)
            figures[var] = {"dist": None, "violin": None}
            figures[var]["dist"] = defaultdict()
            figures[var]["dist"]["figure"], figures[var]["dist"]["axes"] = topo_sum.sns_plot()

            if violin:
                figures[var]["violin"] = defaultdict()
                figures[var]["violin"]["figure"], figures[var]["violin"]["axes"] = topo_sum.sns_violinplot()
        else:
            LOGGER.info(f"[plotting] Statistic is not in dataframe : {var}")
    if pickle_plots:
        outfile = Path(config["output_dir"]) / "distribution_plots.pkl"
        save_pkl(outfile=outfile, to_pkl=figures)
        LOGGER.info(f"[plotting] Images pickled to : {outfile}")

    return figures


def main(args=None):
    """Run Plotting"""

    parser = create_parser()
    args = parser.parse_args() if args is None else parser.parse_args(args)

    if args.config_file is not None:
        config = read_yaml(args.config_file)
        LOGGER.info(f"[plotting] Configuration file loaded from : {args.config_file}")
    else:
        summary_yaml = pkg_resources.open_text(__package__, "summary_config.yaml")
        config = yaml.safe_load(summary_yaml.read())
        LOGGER.info("[plotting] Default configuration file loaded.")
    config = update_config(config, args)
    if args.var_to_label is not None:
        config["var_to_label"] = read_yaml(args.var_to_label)
        LOGGER.info("[plotting] Variable to labels mapping loaded from : {args.var_to_label}")
    else:
        plotting_yaml = pkg_resources.open_text(__package__, "var_to_label.yaml")
        config["var_to_label"] = yaml.safe_load(plotting_yaml.read())
        LOGGER.info("[plotting] Default variable to labels mapping loaded.")
    if args.csv_file is not None:
        config["csv_file"] = args.csv_file

    # Write sample configuration if asked to do so and exit
    if args.create_config_file:
        write_yaml(
            config,
            output_dir="./",
            config_file=args.create_config_file,
            header_message="Sample configuration file auto-generated",
        )
        LOGGER.info(f"A sample configuration has been written to : ./{args.create_config_file}")
        LOGGER.info(
            "Please refer to the documentation on how to use the configuration file : \n\n"
            "https://afm-spm.github.io/TopoStats/usage.html#configuring-topostats\n"
            "https://afm-spm.github.io/TopoStats/configuration.html"
        )
        sys.exit()
    if args.create_label_file:
        write_yaml(
            config["var_to_label"],
            output_dir="./",
            config_file=args.create_label_file,
            header_message="Sample label file auto-generated",
        )
        LOGGER.info(f"A sample label file has been written to : ./{args.create_label_file}")
        LOGGER.info(
            "Please refer to the documentation on how to use the configuration file : \n\n"
            "https://afm-spm.github.io/TopoStats/usage.html#configuring-topostats\n"
            "https://afm-spm.github.io/TopoStats/configuration.html"
        )
        sys.exit()

    # Plot statistics
    toposum(config)


if __name__ == "__main__":
    main()
