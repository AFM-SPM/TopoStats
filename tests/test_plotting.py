"""Tests for the plotting module."""
import importlib.resources as pkg_resources
from pathlib import Path
import yaml

from matplotlib.figure import Figure
import pandas as pd
import pytest

import topostats
from topostats.plotting import TopoSum, toposum
from topostats.plotting import main as plotting_main

# pylint: disable=protected-access

BASE_DIR = Path.cwd()
RESOURCES = BASE_DIR / "tests" / "resources"


def test_df_from_csv(minicircle_all_statistics: pd.DataFrame, toposum_object: TopoSum) -> None:
    """Test loading of CSV file."""
    assert isinstance(toposum_object.df, pd.DataFrame)
    pd.testing.assert_frame_equal(minicircle_all_statistics, toposum_object.df)


def test_toposum_class(toposum_object: TopoSum) -> None:
    """Check the TopoSum class has been correctly instantiated."""
    assert isinstance(toposum_object.df, pd.DataFrame)
    assert isinstance(toposum_object.stat_to_sum, str)
    assert isinstance(toposum_object.molecule_id, str)
    assert isinstance(toposum_object.image_id, str)
    assert isinstance(toposum_object.hist, bool)
    assert isinstance(toposum_object.kde, bool)
    assert isinstance(toposum_object.file_ext, str)
    assert isinstance(toposum_object.output_dir, Path)
    assert isinstance(toposum_object.var_to_label, dict)


def test_outfile(toposum_object: TopoSum) -> None:
    """Check fig and ax returned"""
    outfile = toposum_object._outfile(plot_suffix="testing")
    assert isinstance(outfile, str)
    assert outfile == "area_testing"


def test_args_input_csv() -> None:
    """Test modifying the configuration value for the input CSV to be summarised."""
    assert True


def test_var_to_label_config(tmp_path: Path) -> None:
    """Test the var_to_label configuration file is created correctly."""
    with pytest.raises(SystemExit):
        plotting_main(
            args=[
                "--create-label-file",
                f"{tmp_path / 'var_to_label_config.yaml'}",
                "--input_csv",
                f"{str(RESOURCES / 'minicircle_default_all_statistics.csv')}",
            ]
        )
    var_to_label_config = tmp_path / "var_to_label_config.yaml"
    with var_to_label_config.open("r", encoding="utf-8") as f:
        var_to_label_str = f.read()
    var_to_label = yaml.safe_load(var_to_label_str)
    plotting_yaml = pkg_resources.open_text(topostats.__package__, "var_to_label.yaml")
    expected_var_to_label = yaml.safe_load(plotting_yaml.read())

    assert var_to_label == expected_var_to_label


@pytest.mark.parametrize(
    "var,expected_label",
    [
        ("contour_lengths", "Contour Lengths"),
        ("end_to_end_distance", "End to End Distance"),
        ("grain_bound_len", "Circumference"),
        ("grain_curvature1", "Smaller Curvature"),
    ],
)
def test_set_label(toposum_object: TopoSum, var: str, expected_label: str) -> None:
    """Test labels are returned correctly."""
    toposum_object._set_label(var)
    assert toposum_object.label == expected_label


def test_set_label_keyerror(toposum_object: TopoSum) -> None:
    """Test labels are returned correctly."""
    with pytest.raises(KeyError):
        toposum_object._set_label("non_existent_stat")


def test_toposum(summary_config: dict) -> None:
    """Test the toposum function returns a dictionary with figures."""
    summary_config["csv_file"] = RESOURCES / "minicircle_default_all_statistics.csv"
    summary_config["df"] = pd.read_csv(summary_config["csv_file"])
    summary_config["violin"] = True
    summary_config["stats_to_sum"] = ["area"]
    summary_config["pickle_plots"] = True
    summary_config.pop("stat_to_sum")
    figures = toposum(summary_config)
    assert isinstance(figures, dict)
    assert "area" in figures.keys()
    assert isinstance(figures["area"]["dist"]["figure"], Figure)
    assert isinstance(figures["area"]["violin"]["figure"], Figure)


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_kde(summary_config: dict) -> None:
    """Regression test for sns_plot() with a single KDE."""
    summary_config["hist"] = False
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    fig, _ = _toposum.sns_plot()
    return fig


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_kde_multiple_images(summary_config: dict, toposum_multiple_images: pd.DataFrame) -> None:
    """Regression test for sns_plot() with multiple KDE."""
    summary_config["hist"] = False
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    _toposum.melted_data = toposum_multiple_images
    fig, _ = _toposum.sns_plot()
    return fig


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_hist(summary_config: dict) -> None:
    """Regression test for sns_plot() with a single histogram."""
    summary_config["kde"] = False
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    fig, _ = _toposum.sns_plot()
    return fig


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_hist_multiple_images(summary_config: dict, toposum_multiple_images: pd.DataFrame) -> None:
    """Regression test for sns_plot() with multiple overlaid histograms."""
    summary_config["kde"] = False
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    _toposum.melted_data = toposum_multiple_images
    fig, _ = _toposum.sns_plot()
    return fig


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_hist_kde(summary_config: dict) -> None:
    """Test plotting Kernel Density Estimate and Histogram for area."""
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    fig, _ = _toposum.sns_plot()
    return fig


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_hist_kde_multiple_images(summary_config: dict, toposum_multiple_images: pd.DataFrame) -> None:
    """Test plotting Kernel Density Estimate and Histogram for area."""
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    _toposum.melted_data = toposum_multiple_images
    fig, _ = _toposum.sns_plot()
    return fig


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_violin(summary_config: dict) -> None:
    """Test plotting Kernel Density Estimate and Histogram for area for a single image."""
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    fig, _ = _toposum.sns_violinplot()
    return fig


@pytest.mark.mpl_image_compare(baseline_dir="resources/img/distributions/")
def test_plot_violin_multiple_images(summary_config: dict, toposum_multiple_images: pd.DataFrame) -> None:
    """Test plotting Kernel Density Estimate and Histogram for area with multiple images."""
    _toposum = TopoSum(csv_file=RESOURCES / "minicircle_default_all_statistics.csv", **summary_config)
    _toposum.melted_data = toposum_multiple_images
    fig, _ = _toposum.sns_violinplot()
    return fig
