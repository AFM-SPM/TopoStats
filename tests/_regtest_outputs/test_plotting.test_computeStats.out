     aspect_ratio          area        volume     min_feret     max_feret  Contour Lengths  End to End Distance
max             0  1.000000e+02  1.000000e+02  1.000000e+02  1.000000e+02     1.000000e+02         1.000000e+02
std             0  3.073343e+10  7.426329e+10  4.527678e+09  3.909266e+09     1.785414e+10         1.149846e+10
ste             0  6.872204e+09  1.660578e+10  1.012420e+09  8.741385e+08     3.992306e+09         2.571133e+09
N               0  2.100000e+01  2.100000e+01  2.100000e+01  2.100000e+01     2.100000e+01         2.100000e+01
